### PR TITLE
[SMTChecker] Uninterpreted functions for blockhash()

### PR DESCRIPTION
**Pull request by leonardoalt**
Depends on #5398 

Uninterpreted functions should also be used for crypto&math, abi.

### CODE REVIEW

File: libsolidity/formal/SMTLib2Interface.cpp [hunk 1]
```diff
@@ -64,17 +64,20 @@ void SMTLib2Interface::pop()
 	m_accumulatedOutput.pop_back();
 }
 
-void SMTLib2Interface::declareFunction(string _name, Sort _domain, Sort _codomain)
+void SMTLib2Interface::declareFunction(string _name, vector<Sort> const& _domain, Sort _codomain)
 {
 	// TODO Use domain and codomain as key as well
+	string domain("");
+	for (auto const& _sort: _domain)
+		domain += toSSort(_sort) + ' ';
```
Comment (ekpyron) [inline libsolidity/formal/SMTLib2Interface.cpp]: Naming the loop variable ``_sort`` made me look for it in the function header...

**Review by ekpyron — DISMISSED** [review]
I'm still getting familiar with the SMT code, so I may be missing stuff, but this looks fine to me,

File: libsolidity/formal/SMTLib2Interface.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTLib2Interface.cpp]: True, I'll change.

**Comment by leonardoalt** [comment]
So, some actual description to help out a bit:
Before, functions were written as 1 argument 1 return (because they were used as the SSA variables). This PR:
- changes that to an arbitrary number of function arguments;
- encodes `blockhash(...)` as an uninterpreted function (UF)
- keeps track of the used UFs and shows the model properly later on

**Comment by ekpyron** [comment]
Actually, do you think it may make sense to make ``Sort`` a proper class that can be an ``IntegerSort`` or a ``BoolSort`` or a ``FunctionSort`` (with domain and codomain members)? Then you could, e.g., replace ``declareBool``, ``declareInteger`` and ``declareFunction`` by ``declare(name, sort)``, etc..
And at some point you may want to represent ~~multiple return values and~~ functions returning functions, etc...
EDIT: multiple return values probably won't work well that way... but functions returning functions might - although the solvers probably won't be able to handle that all that well, but still...

**Comment by leonardoalt** [comment]
Yes, there's another PR with that kind of changes ;)

**Comment by leonardoalt** [comment]
Updated & rebased

**Review by ekpyron — DISMISSED** [review]

File: libsolidity/formal/SMTChecker.h [hunk 2]
```diff
@@ -191,6 +192,11 @@ class SMTChecker: private ASTConstVisitor
 	std::unordered_map<Expression const*, std::shared_ptr<SymbolicVariable>> m_expressions;
 	std::unordered_map<VariableDeclaration const*, std::shared_ptr<SymbolicVariable>> m_variables;
 	std::unordered_map<std::string, std::shared_ptr<SymbolicVariable>> m_specialVariables;
+	/// Stores the declaration of an Uninterpreted Function.
+	std::unordered_map<std::string, smt::Expression> m_uFunctions;
+	/// Stores the instances of an UF applied to arguments.
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.h]:
```suggestion
	/// Stores the instances of an Uninterpreted Function applied to arguments.
```

File: libsolidity/formal/SMTChecker.h [hunk 3]
```diff
@@ -191,6 +192,11 @@ class SMTChecker: private ASTConstVisitor
 	std::unordered_map<Expression const*, std::shared_ptr<SymbolicVariable>> m_expressions;
 	std::unordered_map<VariableDeclaration const*, std::shared_ptr<SymbolicVariable>> m_variables;
 	std::unordered_map<std::string, std::shared_ptr<SymbolicVariable>> m_specialVariables;
+	/// Stores the declaration of an Uninterpreted Function.
+	std::unordered_map<std::string, smt::Expression> m_uFunctions;
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.h]:
```suggestion
	std::unordered_map<std::string, smt::Expression> m_uninterpretedFunctions;
```

File: libsolidity/formal/SMTChecker.h [hunk 4]
```diff
@@ -191,6 +192,11 @@ class SMTChecker: private ASTConstVisitor
...
+	/// Stores the instances of an UF applied to arguments.
+	/// Used to retrieve models.
+	std::vector<smt::Expression> m_uTerms;
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.h]:
```suggestion
	std::vector<smt::Expression> m_uninterpretedTerms;
```

File: libsolidity/formal/SMTChecker.cpp [hunk 5]
```diff
@@ -788,11 +812,33 @@ void SMTChecker::checkCondition(
...
 			map<string, string> sortedModel;
-			for (size_t i = 0; i < values.size(); ++i)
+			// Handle program variables first.
+			for (size_t i = 0; i < expressionNames.size(); ++i)
 				if (expressionsToEvaluate.at(i).name != values.at(i))
 					sortedModel[expressionNames.at(i)] = values.at(i);
+			// Handle uninterpreted functions.
+			for (size_t i = expressionNames.size(); i < values.size() - m_expressions.size(); ++i)
+			{
+				auto const& uf = expressionsToEvaluate.at(i);
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]:
```suggestion
				smt::Expression const& uf = expressionsToEvaluate.at(i);
```

File: libsolidity/formal/SMTLib2Interface.h [hunk 6]
```diff
@@ -58,6 +58,7 @@ class SMTLib2Interface: public SolverInterface, public boost::noncopyable
 
 private:
 	std::string toSExpr(Expression const& _expr);
+	std::string toSSort(Sort _sort);
```
Comment (chriseth) [inline libsolidity/formal/SMTLib2Interface.h]: What does the `S` stand for?

File: libsolidity/formal/SMTChecker.cpp [hunk 7]
```diff
@@ -768,6 +781,17 @@ void SMTChecker::checkCondition(
 			expressionsToEvaluate.emplace_back(var.second->currentValue());
 			expressionNames.push_back(var.first);
 		}
+		// Names for UFs are created later.
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: This is getting rather complicated and should probably be split into multiple functions.

**Review by chriseth — DISMISSED** [review]

File: libsolidity/formal/SMTLib2Interface.h (hunk 6 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTLib2Interface.h]: Updated to `SmtLibSort`

**Comment by leonardoalt** [comment]
Updated

**Review by axic — CHANGES_REQUESTED** [review]
Please rebase and move the changelog entry.

(Also collapse the addressed comments.)

**Comment by leonardoalt** [comment]
@axic updated

File: Changelog.md [hunk 8]
```diff
@@ -5,6 +5,8 @@ Language Features:
 
 Compiler Features:
  * Build System: LLL is not built anymore by default. Must configure it with CMake as `-DLLL=ON`.
+ * SMTChecker: Support msg, tx and block member variables.
+ * SMTChecker: Support gasleft() and blockhash() functions.
```
Comment (axic) [inline Changelog.md]: Aw, these should be escaped for pretty printing :)

File: libsolidity/formal/SMTChecker.cpp [hunk 9]
```diff
@@ -889,7 +906,8 @@ SMTChecker::checkSatisfiableAndGenerateModel(vector<smt::Expression> const& _exp
 		try
 		{
 			// Parse and re-format nicely
-			value = formatNumber(bigint(value));
+			if (isValidDecimal(value) || isValidHex(value))
```
Comment (axic) [inline libsolidity/formal/SMTChecker.cpp]: This seems to be an unrelated change?

If so would split it into a separate PR or at least separate commit.

**Comment by axic** [comment]
I haven't done a full review yet.

File: libsolidity/formal/SMTChecker.cpp (hunk 9 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp]: Removed

File: libsolidity/formal/SMTChecker.cpp [hunk 10]
```diff
@@ -78,6 +78,9 @@ bool SMTChecker::visit(FunctionDefinition const& _function)
 		m_interface->reset();
 		m_pathConditions.clear();
 		m_expressions.clear();
+		m_specialVariables.clear();
```
Comment (axic) [inline libsolidity/formal/SMTChecker.cpp:81]: Hm, is this a bugfix? Or it just wasn't used before?

**Review by axic — APPROVED** [review]
Apart from that single question it looks good.

File: libsolidity/formal/SMTChecker.cpp (hunk 10 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp:81]: The symbolic variables used by the special variables can live across transactions, that's why this wasn't there before, but I think it's safer to clear them anyway.
